### DISCUSSION TITLE
[DropdownMenu] How can i test keyboard navigation?

### DISCUSSION THREAD

**Discussion by Luccasoli** [radix-ui/primitives#1063] [Help]
I'm using react-testing-library, @testing-library/dom, @testing-library/user-event and jest

i tried this:

```js
    render(
      <Menu>
        <MenuButton>
          <Box role="button">Test Menu</Box>
        </MenuButton>

        <MenuList>
          <MenuItem>
            <MenuLabel>Item 1</MenuLabel>
          </MenuItem>
          <MenuItem>
            <MenuLabel>Item 2</MenuLabel>
          </MenuItem>
          <MenuItem>
            <MenuLabel>Item 3</MenuLabel>
          </MenuItem>
        </MenuList>
      </Menu>,
    )

    const menuButton = screen.getByRole('button')
    userEvent.click(menuButton)

    const menu = screen.getByRole('menu')
    fireEvent.keyDown(menu, {key: 'ArrowDown'})

    const menuItem1 = screen.getByRole('menuitem', {name: 'Item 1'})
    expect(menuItem1).toHaveFocus()

    const menuItem2 = screen.getByRole('menuitem', {name: 'Item 2'})
    fireEvent.keyDown(menuItem1, {
      key: 'ArrowDown',
    })
    expect(menuItem2).toHaveFocus()
```

But my second item never gets focused :(

Error: 
![image](https://user-images.githubusercontent.com/31039396/147950274-a6e73368-669d-471f-aa96-ca54e6b8d1af.png)

under the hood:
Menu = Root
MenuButton = MenuTrigger with asChild
MenuList = Content
MenuItem = Item

**Comment by jjenzz** [comment]
It's not immediately obvious (to me) why your code isn't working without setting up the same environment as you so hopefully someone else has some ideas. However, it looks a lot like you are trying to test our functionality rather than your own which should be redundant. Ideally, we would have tests for all of that [here](https://github.com/radix-ui/primitives/blob/main/cypress/integration/DropdownMenu.spec.ts) so that you don't have to.

If anyone is open to contributing more tests to our spec, that would be super appreciated.

**Reply by Luccasoli** [reply]
Thanks for answering, @jjenzz!

My tech lead told me the same about re-test 😂 but the idea is to have this tests in order to ensure the functionilities keep working even if we refactor replacing/removing radix-ui from source (which is something that i dont see happening)

And i saw [this tests](https://github.com/radix-ui/primitives/blob/main/cypress/integration/DropdownMenu.spec.ts), but i've not seen that navigation that i'm trying to accomplish (vertical navigation with up/down arrows or with home/end keys), if i resolve this test problem I should certainly open a PR by adding this test in the radixui/primitives source code

**Reply by jjenzz** [reply]
The issue you are having is likely easier to solve in our tests since we use cypress and not jest. 

> but the idea is to have this tests in order to ensure the functionilities keep working even if we refactor replacing/removing radix-ui from source (which is something that i dont see happening)

If you don't see it happening then it is likely a premature task. Perhaps easier to add tests for these things **if/when** you decide to customise the underlying functionality? Mostly because, if the underlying functionality is changed to another library instead for example, you would still be "re-testing".

**Reply by Luccasoli** [reply]
> The issue you are having is likely easier to solve in our tests since we use cypress and not jest.

Yes! For now we still using jest, but maybe it's something we could use it too for unit tests!

> If you don't see it happening then it is likely a premature task. Perhaps easier to add tests for these things if/when you decide to customise the underlying functionality? Mostly because, if the underlying functionality is changed to another library instead for example, you would still be "re-testing".

i see 🤔 i thought if i test just using this a11y roles this tests could be reused in different implementations/libraries, and yes i will just skip this test for now, but i really appreciate you sharing your thoughts about it! Thanks!

**Comment by penx** [comment]
I understand the point about not needing to test the radix functionality in unit tests, but our use case would be:

- integration tests that rely on the outcome of clicking on a submenu item
- visual regression tests that ensure the submenu is mounted _after_ the main menu
- confidence when upgrading to new major versions of radix that could have API changes
- quickly opening our storybook story in a specific state during development

We're using storybook play functions with @storybook/jest and @storybook/testing-library for this. 

I tried

- userEvent.hover
- fireEvent.pointerOver
- fireEvent.keyDown

And initially ran in to similar issues, but have solved it by making sure the event is sent to the currently focussed item:


```ts
  // in Storybook, in order to query the portal, we need to access the parent element
  const canvas = within(canvasElement.parentElement!);

  await waitFor(() => {
    expect(canvas.getByRole("menuitem", { name: "Third Item" })).toBeVisible();
  });

  fireEvent.keyDown(canvas.getByRole("menu"), {
    key: "ArrowDown",
    code: "ArrowDown",
    charCode: 40,
  });

  await waitFor(() =>
    expect(
      canvas.getByRole("menuitem", {
        name: "First Item",
      }),
    ).toHaveFocus(),
  );
  fireEvent.keyDown(canvas.getByRole("menuitem", { name: "First Item" }), {
    key: "ArrowDown",
    code: "ArrowDown",
    charCode: 40,
  });

  await waitFor(() =>
    expect(
      canvas.getByRole("menuitem", {
        name: "Second Item",
      }),
    ).toHaveFocus(),
  );
  fireEvent.keyDown(canvas.getByRole("menuitem", { name: "Second Item" }), {
    key: "ArrowDown",
    code: "ArrowDown",
    charCode: 40,
  });

  await waitFor(() =>
    expect(
      canvas.getByRole("menuitem", {
        name: "Third Item",
      }),
    ).toHaveFocus(),
  );
  fireEvent.keyDown(canvas.getByRole("menuitem", { name: "Third Item" }), {
    key: "ArrowRight",
    code: "ArrowRight",
    charCode: 39,
  });

  await waitFor(() => {
    expect(canvas.getByText("Submenu Item")).toBeVisible();
  });
```
